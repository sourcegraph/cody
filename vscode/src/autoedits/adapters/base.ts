import type { PromptString } from '@sourcegraph/cody-shared'

export interface AutoeditsModelAdapter {
    getModelResponse(args: AutoeditModelOptions): Promise<string>
}

/**
 * Represents the structure of a prompt for auto-edits functionality
 */
export type AutoeditsPrompt = {
    /**
     * Optional system message to provide context or instructions
     * This field is only valid for the chat models.
     * For the completions models, this is ignored by the adapters.
     */
    systemMessage?: PromptString
    /**
     * The user message containing the code to be rewritten.
     */
    userMessage: PromptString
}

export interface AutoeditModelOptions {
    url: string
    model: string
    apiKey: string
    prompt: AutoeditsPrompt
    codeToRewrite: string
    userId: string | null
    isChatModel: boolean
}
