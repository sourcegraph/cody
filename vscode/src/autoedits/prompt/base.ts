import type { AutoEditsTokenLimit, PromptString } from '@sourcegraph/cody-shared'
import type {
    AutocompleteContextSnippet,
    DocumentContext,
} from '@sourcegraph/cody-shared/src/completions/types'
import type * as vscode from 'vscode'
import type { CodeToReplaceData } from './prompt-utils'

export interface UserPromptArgs {
    docContext: DocumentContext
    document: vscode.TextDocument
    position: vscode.Position
    context: AutocompleteContextSnippet[]
    tokenBudget: AutoEditsTokenLimit
}

export interface UserPromptResponse {
    codeToReplace: CodeToReplaceData
    prompt: PromptString
}

/**
 * Interface for generating user prompts in auto-edits functionality.
 * The major difference between different strategy is the prompt rendering.
 */
export interface AutoeditsUserPromptStrategy {
    /**
     * Generates a prompt string based on the provided arguments.
     * @param args - The arguments containing document context, position, and token budget.
     * @returns A promise that resolves to a prompt string.
     */
    getUserPrompt(args: UserPromptArgs): UserPromptResponse
}
