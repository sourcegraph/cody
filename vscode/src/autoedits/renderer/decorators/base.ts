import type * as vscode from 'vscode'

/**
 * Represents a decorator that manages VS Code editor decorations for auto-edit suggestions.
 *
 * This interface defines the contract for displaying and managing decorative elements
 * that visualize proposed text changes in the editor.
 *
 * Lifecycle:
 * - A single instance should be created per decoration session and disposed of when the decorations
 *   are no longer needed.
 * - Always call dispose() when the decorator is no longer needed to clean up resources.
 * - Dispose should always clear the decorations.
 *
 * Usage Pattern:
 * ```typescript
 * const decorator = createAutoEditsDecorator(...);
 * try {
 *   decorator.setDecorations(decorationInfo);
 *   ...
 * } finally {
 *   decorator.dispose();
 * }
 * ```
 */
export interface AutoEditsDecorator extends vscode.Disposable {
    /**
     * Applies decorations to the editor based on the provided decoration information.
     *
     * @param decorationInfo Contains the line-by-line information about text changes
     *        and how they should be decorated in the editor.
     */
    setDecorations(decorationInfo: DecorationInfo): void
}

/**
 * Represents a line of text with its change type and content.
 */
export type DecorationLineInfo = AddedLineInfo | RemovedLineInfo | ModifiedLineInfo | UnchangedLineInfo

export interface AddedLineInfo {
    id: string
    type: 'added'
    text: string
    modifiedLineNumber: number
}

export interface RemovedLineInfo {
    id: string
    type: 'removed'
    text: string
    originalLineNumber: number
}

export interface ModifiedLineInfo {
    id: string
    type: 'modified'
    oldText: string
    newText: string
    changes: LineChange[]
    originalLineNumber: number
    modifiedLineNumber: number
}

export interface UnchangedLineInfo {
    id: string
    type: 'unchanged'
    text: string
    originalLineNumber: number
    modifiedLineNumber: number
}

export type LineChange = {
    id: string
    type: 'insert' | 'delete' | 'unchanged'
    /** `range` in the original text relative to the document start */
    originalRange: vscode.Range
    /** `range` in the modified text relative to the document start */
    modifiedRange: vscode.Range
    text: string
}

export interface DecorationInfo {
    modifiedLines: ModifiedLineInfo[]
    removedLines: RemovedLineInfo[]
    addedLines: AddedLineInfo[]
    unchangedLines: UnchangedLineInfo[]
}
