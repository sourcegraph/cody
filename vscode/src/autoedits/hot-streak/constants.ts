/**
 * Minimum number of lines that should be accumulated before attempting a hot streak suggestion.
 * Note: Reaching this number does not guarantee a hot streak suggestion will be emitted.
 * The suggestion should also produce a suitable diff that we can present to the user.
 */
export const SHOULD_ATTEMPT_HOT_STREAK_CHUNK_THRESHOLD = 5

/**
 * Minimum number of lines that should be present in a hot-streak chunk before it is suggested to the user.
 * This is to avoid suggesting a hot-streak chunk that is too small and atomic.
 *
 * Note: This differs from `SHOULD_ATTEMPT_HOT_STREAK_CHUNK_THRESHOLD` in that the chunk is trimmed based on the
 * first stable "unchanged" area. This means that, even if the prediction meets this threshold, the actual chunk may
 * be smaller.
 *
 * Example for the above scenario:
 * 1. LLM produces a response that is 5 lines long, we deem it suitable to attempt a hot-streak chunk.
 * 2. Our diff logic finds that the first line changed, then we had 3 unchanged lines, then 1 changed line.
 * 3. To ensure we can reliably use the hot-streak chunk, our diff logic trims to chunk to the first unchanged area.
 *    This results in a total chunk length of 4.
 * 4. We don't want to use this chunk as it is too small.
 */
export const SHOULD_USE_HOT_STREAK_CHUNK_THRESHOLD = 5

/**
 * The amount of lines that should be present in a stable "unchanged" hunk before it is considered stable.
 * This is to help avoid false positives where the LLM produces new code which we similar to existing code below that
 * we mark as stable.
 *
 * Note: This can be tweaked to improve the stability of the partial diff.
 */
export const SHOULD_USE_STABLE_UNCHANGED_HUNK_THRESHOLD = 2
