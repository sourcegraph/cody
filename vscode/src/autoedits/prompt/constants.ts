import { ps } from '@sourcegraph/cody-shared'

export const LINT_ERRORS_TAG_OPEN = ps`<lint_errors>`
export const LINT_ERRORS_TAG_CLOSE = ps`</lint_errors>`
export const EXTRACTED_CODE_SNIPPETS_TAG_OPEN = ps`<extracted_code_snippets>`
export const EXTRACTED_CODE_SNIPPETS_TAG_CLOSE = ps`</extracted_code_snippets>`
export const SNIPPET_TAG_OPEN = ps`<snippet>`
export const SNIPPET_TAG_CLOSE = ps`</snippet>`
export const RECENT_SNIPPET_VIEWS_TAG_OPEN = ps`<recently_viewed_snippets>`
export const RECENT_SNIPPET_VIEWS_TAG_CLOSE = ps`</recently_viewed_snippets>`
export const RECENT_EDITS_TAG_OPEN = ps`<diff_history>`
export const RECENT_EDITS_TAG_CLOSE = ps`</diff_history>`
export const RECENT_COPY_TAG_OPEN = ps`<recent_copy>`
export const RECENT_COPY_TAG_CLOSE = ps`</recent_copy>`
export const FILE_TAG_OPEN = ps`<file>`
export const FILE_TAG_CLOSE = ps`</file>`
export const AREA_FOR_CODE_MARKER = ps`<<<AREA_AROUND_CODE_TO_REWRITE_WILL_BE_INSERTED_HERE>>>`
export const AREA_FOR_CODE_MARKER_OPEN = ps`<area_around_code_to_rewrite>`
export const AREA_FOR_CODE_MARKER_CLOSE = ps`</area_around_code_to_rewrite>`
export const CODE_TO_REWRITE_TAG_CLOSE = ps`</code_to_rewrite>`
export const CODE_TO_REWRITE_TAG_OPEN = ps`<code_to_rewrite>`

// Some common prompt instructions
export const SYSTEM_PROMPT = ps`You are an intelligent programmer named CodyBot. You are an expert at coding. Your goal is to help your colleague finish a code change.`
export const BASE_USER_PROMPT = ps`Help me finish a coding change. In particular, you will see a series of snippets from current open files in my editor, files I have recently viewed, the file I am editing, then a history of my recent codebase changes, then current compiler and linter errors, content I copied from my codebase. You will then rewrite the <code_to_rewrite>, to match what you think I would do next in the codebase. Note: I might have stopped in the middle of typing.`
export const FINAL_USER_PROMPT = ps`Now, continue where I left off and finish my change by rewriting "code_to_rewrite":`
export const LONG_TERM_SNIPPET_VIEWS_INSTRUCTION = ps`Here are some snippets of code I have recently viewed, roughly from oldest to newest. It's possible these aren't entirely relevant to my code change:`
export const JACCARD_SIMILARITY_INSTRUCTION = ps`Here are some snippets of code I have extracted from open files in my code editor. It's possible these aren't entirely relevant to my code change:`
export const RECENT_EDITS_INSTRUCTION = ps`Here is my recent series of edits from oldest to newest.`
export const LINT_ERRORS_INSTRUCTION = ps`Here are some linter errors from the code that you will rewrite.`
export const RECENT_COPY_INSTRUCTION = ps`Here is some recent code I copied from the editor.`
export const CURRENT_FILE_INSTRUCTION = ps`Here is the file that I am looking at `
export const SHORT_TERM_SNIPPET_VIEWS_INSTRUCTION = ps`Here are some snippets of code I just looked at:`
