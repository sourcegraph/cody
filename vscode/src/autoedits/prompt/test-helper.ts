import dedent from 'dedent'
import { getCurrentDocContext } from '../../completions/get-current-doc-context'
import { documentAndPosition } from '../../completions/test-helpers'
import { type CodeToReplaceData, getCurrentFilePromptComponents } from '../prompt/prompt-utils'

interface CodeToReplaceTestOptions {
    maxPrefixLength: number
    maxSuffixLength: number
    maxPrefixLinesInArea: number
    maxSuffixLinesInArea: number
    codeToRewritePrefixLines: number
    codeToRewriteSuffixLines: number
}

export function createCodeToReplaceDataForTest(
    code: TemplateStringsArray,
    options: CodeToReplaceTestOptions,
    ...values: unknown[]
): CodeToReplaceData {
    const { document, position } = documentAndPosition(dedent(code, values))
    const docContext = getCurrentDocContext({
        document,
        position,
        maxPrefixLength: options.maxPrefixLength,
        maxSuffixLength: options.maxSuffixLength,
    })

    return getCurrentFilePromptComponents({
        docContext,
        position,
        document,
        tokenBudget: options,
    }).codeToReplaceData
}
