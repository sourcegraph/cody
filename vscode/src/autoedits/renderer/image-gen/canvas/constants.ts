import type { DiffColors } from './types'

export const DIFF_COLORS = {
    inserted: {
        line: 'rgba(155, 185, 85, 0.1)',
        text: 'rgba(155, 185, 85, 0.15)',
    },
    removed: {
        line: 'rgba(255, 0, 0, 0.1)',
        text: 'rgba(255, 0, 0, 0.15)',
    },
} satisfies DiffColors
