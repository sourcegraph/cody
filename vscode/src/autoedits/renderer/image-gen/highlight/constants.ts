import type { SYNTAX_HIGHLIGHT_THEME } from './types'

export const DEFAULT_HIGHLIGHT_COLORS: Record<SYNTAX_HIGHLIGHT_THEME, string> = {
    dark: '#ffffff',
    light: '#000000',
}
