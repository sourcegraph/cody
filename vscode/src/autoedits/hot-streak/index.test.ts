import dedent from 'dedent'
import { describe, expect, it } from 'vitest'
import { type ProcessHotStreakResponsesParams, processHotStreakResponses } from '.'
import { getCurrentDocContext } from '../../completions/get-current-doc-context'
import { documentAndPosition } from '../../completions/test-helpers'
import { AutoeditStopReason, type ModelResponse } from '../adapters/base'
import type { SuggestedPredictionResult } from '../autoedits-provider'
import { LONG_SUGGESTION_USER_CURSOR_MARKER } from '../prompt/constants'
import { createCodeToReplaceDataForTest } from '../prompt/test-helper'

// Helper to create a generator for model responses
async function* createModelResponseGenerator(
    predictions: string[],
    type: 'partial' | 'success' | 'aborted' = 'partial',
    stopReason = AutoeditStopReason.RequestFinished
): AsyncGenerator<ModelResponse> {
    let cumulativePrediction = ''

    // Yield all but the last prediction as partial responses
    for (let i = 0; i < predictions.length - 1; i++) {
        cumulativePrediction += predictions[i] + '\n'
        yield {
            type: 'partial',
            prediction: cumulativePrediction,
            stopReason: AutoeditStopReason.StreamingChunk,
            requestHeaders: {},
            requestUrl: 'test-url',
            responseHeaders: {},
            responseBody: {},
        }
    }

    if (predictions.length > 0) {
        cumulativePrediction += predictions[predictions.length - 1]
        yield {
            type: 'success',
            prediction: cumulativePrediction,
            stopReason: AutoeditStopReason.RequestFinished,
            responseBody: {},
            requestHeaders: {},
            responseHeaders: {},
            requestUrl: 'test-url',
        }
    }
}

const MOCK_CODE = `
export function isEvenOrOdd(numberToChange: number): boolean {█
    // Check if target is 0
    if (numberToChange === 0) {
        return true
    }
    // Check if target is 1
    if (numberToChange === 1) {
        return false
    }
    // Check if target is 2
    if (numberToChange === 2) {
        return true
    }
    // Check if target is 3
    if (numberToChange === 3) {
        return false
    }
    // Check if target is 4
    if (numberToChange === 4) {
        return true
    }
    // Check if target is 5
    if (numberToChange === 5) {
        return false
    }
    throw new Error('Out of RAM')
}
`

const MOCK_CODE_WITHOUT_CURSOR = MOCK_CODE.replace('█', '')

const MOCK_PREDICTION = `
export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }
    if (target === 4) {
        return true
    }
    if (target === 5) {
        return false
    }
    throw new Error('Out of RAM')
}
`

function createHotStreakParams(
    code: string,
    responseGenerator: AsyncGenerator<ModelResponse>
): ProcessHotStreakResponsesParams {
    const { document, position } = documentAndPosition(code)
    const codeToReplaceData = createCodeToReplaceDataForTest(code, {
        maxPrefixLength: 2000,
        maxSuffixLength: 2000,
        maxPrefixLinesInArea: 2,
        maxSuffixLinesInArea: 2,
        codeToRewritePrefixLines: 1,
        codeToRewriteSuffixLines: 30,
        prefixTokens: 100,
        suffixTokens: 100,
    })
    const docContext = getCurrentDocContext({
        document,
        position,
        maxPrefixLength: 2000,
        maxSuffixLength: 2000,
    })

    return {
        responseGenerator,
        document,
        codeToReplaceData,
        requestDocContext: docContext,
        position,
        options: {
            hotStreakEnabled: true,
        },
    }
}

describe('processHotStreakResponses', () => {
    it('post-processes predictions to filter out marker tokens', async () => {
        // Create prediction with the marker token that should be filtered out
        const markerToken = LONG_SUGGESTION_USER_CURSOR_MARKER
        const predictionWithMarker = MOCK_PREDICTION.replace('return true', `${markerToken}return true`)
        const responseGenerator = createModelResponseGenerator([predictionWithMarker])
        const params = createHotStreakParams(MOCK_CODE, responseGenerator)
        const resultGenerator = processHotStreakResponses(params)

        // Collect all results
        const results = []
        for await (const result of resultGenerator) {
            results.push(result)
        }

        // Verify the marker has been filtered out in all results
        for (const result of results) {
            if (result.type !== 'aborted') {
                expect(result.response.prediction).not.toContain(markerToken)
            }
        }

        // Verify we got at least one result
        expect(results.length).toBeGreaterThan(0)
    })

    it('does not emit hot streaks if disabled', async () => {
        const responseGenerator = createModelResponseGenerator(MOCK_PREDICTION.split('\n'))
        const params = createHotStreakParams(MOCK_CODE, responseGenerator)
        const resultGenerator = processHotStreakResponses({
            ...params,
            options: { hotStreakEnabled: false },
        })

        const results = []
        for await (const result of resultGenerator) {
            results.push(result)
        }
        expect(results.length).toBe(1)
        const finalResponse = results[0] as SuggestedPredictionResult
        expect(finalResponse.type).toBe('suggested')
        expect(finalResponse.response.prediction).toBe(MOCK_PREDICTION)
        expect(finalResponse.response.stopReason).toBe(AutoeditStopReason.RequestFinished) // No hot-streak
        expect(finalResponse.codeToReplaceData.codeToRewrite).toBe(MOCK_CODE_WITHOUT_CURSOR)
    })

    it('does emit hot streaks when enabled', async () => {
        const responseGenerator = createModelResponseGenerator(MOCK_PREDICTION.split('\n'))
        const params = createHotStreakParams(MOCK_CODE, responseGenerator)
        const resultGenerator = processHotStreakResponses(params)

        const results = []
        for await (const result of resultGenerator) {
            results.push(result)
        }

        const suggestedResults = results.filter(result => result.type === 'suggested')
        expect(suggestedResults.length).toBe(6)

        let resultSnapshot = ''
        for (const result of suggestedResults) {
            expect(result.type).toBe('suggested')
            expect(result.response.stopReason).toBe(AutoeditStopReason.HotStreak)
            resultSnapshot += `
Code to Rewrite:
${result.codeToReplaceData.codeToRewrite}
Prediction:
${result.response.prediction}
            `
        }
        expect(resultSnapshot).toMatchSnapshot()
    })

    it('does not emit hot streak if its suffix is already in the code', async () => {
        const documentText = dedent`import argparse

            def run_batch_job(sanitized_mu, sanitized_dn, sanitized_ts, sanitized_n, sanitized_h, sanitized_pid):
                # Placeholder function for running the batch job
                print(f"Running batch job with:\nModel URL: {sanitized_mu}\nDataset Name: {sanitized_dn}\n"
                    f"Training Script: {sanitized_ts}\nIterations: {sanitized_n}\n"
                    f"Hyperparameters: {sanitized_h}\nProject ID: {sanitized_pid}")

            def main():
                parser = argparse.ArgumentParser(description='Run batch job process.')
                parser.add_argument('--use_cached_model', action='store_true', he█)
                parser.add_argument('--num_iterations', type=int)
                parser.add_argument('--model_url')
                parser.add_argument('--dataset_name')
                parser.add_argument('--training_script')
                parser.add_argument('--hyperparameters')
                parser.add_argument('--project_id')

                args = parser.parse_args()

                sanitized_mu = args.model_url or input("Enter the model URL: ")
                sanitized_dn = args.dataset_name or input("Enter the dataset name: ")
                sanitized_ts = args.training_script or input("Enter the path to the training script: ")
                sanitized_h = args.hyperparameters or input("Enter the hyperparameters for batch job (in JSON format): ")
                sanitized_n = args.num_iterations or input("Enter the number of iterations: ")
                sanitized_pid = args.project_id or input("Enter the project ID: ")

                run_batch_job(sanitized_mu, sanitized_dn, sanitized_ts, sanitized_n, sanitized_h, sanitized_pid)

            if __name__ == '__main__':
                main()\n`

        // The second hot streak prediction ends with the last line of the document code.
        // But the codeToReplaceData.range.end is at the start of the last line, which means
        // the decoration info will consider the last line to be modified with the last line
        // context insertion.
        //
        // This creates a hot-streak item that we later will move the cursor to. And this item
        // will be hidden by the autoedit-provider because the end of the prediction matches
        // text that is already in the document.
        //
        // To fix that we can check for the suffix match before emiting the hot-streak item.
        const predictionText = dedent`
                parser = argparse.ArgumentParser(description='Run batch job process.')
                parser.add_argument('--use_cached_model', action='store_true', help='Use cached model')
                parser.add_argument('--num_iterations', type=int, help='Number of iterations')
                parser.add_argument('--model_url', help='Model URL')
                parser.add_argument('--dataset_name', help='Dataset name')
                parser.add_argument('--training_script', help='Path to training script')
                parser.add_argument('--hyperparameters', help='Hyperparameters for batch job (in JSON format)')
                parser.add_argument('--project_id', help='Project ID')

                args = parser.parse_args()

            █    sanitized_mu = args.model_url or input("Enter the model URL: ")
                sanitized_dn = args.dataset_name or input("Enter the dataset name: ")
                sanitized_ts = args.training_script or input("Enter the path to the training script: ")
                sanitized_h = args.hyperparameters or input("Enter the hyperparameters for batch job (in JSON format): ")
                sanitized_n = args.num_iterations or input("Enter the number of iterations: ")
                sanitized_pid = args.project_id or input("Enter the project ID: ")

                run_batch_job(sanitized_mu, sanitized_dn, sanitized_ts, sanitized_n, sanitized_h, sanitized_pid)

            if __name__ == '__main__':
                main()`

        const responseGenerator = async function* (
            predictions: string[]
        ): AsyncGenerator<ModelResponse> {
            let cumulativePrediction = ''

            // Yield all but the last prediction as partial responses
            for (let i = 0; i < predictions.length - 1; i++) {
                cumulativePrediction += predictions[i]
                yield {
                    type: 'partial',
                    prediction: cumulativePrediction,
                    stopReason: AutoeditStopReason.StreamingChunk,
                    requestHeaders: {},
                    requestUrl: 'test-url',
                    responseHeaders: {},
                    responseBody: {},
                }
            }

            if (predictions.length > 0) {
                cumulativePrediction += predictions[predictions.length - 1]
                yield {
                    type: 'success',
                    prediction: cumulativePrediction,
                    stopReason: AutoeditStopReason.RequestFinished,
                    responseBody: {},
                    requestHeaders: {},
                    responseHeaders: {},
                    requestUrl: 'test-url',
                }
            }
        }
        const params = createHotStreakParams(documentText, responseGenerator(predictionText.split('█')))
        const resultGenerator = processHotStreakResponses(params)

        const results = []
        for await (const result of resultGenerator) {
            results.push(result)
        }

        // Only the first change is returned, the rest is unchanged
        expect(results.length).toBe(2)

        const firstResponse = results[0] as SuggestedPredictionResult
        expect(firstResponse.type).toBe('suggested')
        expect(firstResponse.response.prediction).toMatchInlineSnapshot(`
          "parser = argparse.ArgumentParser(description='Run batch job process.')
              parser.add_argument('--use_cached_model', action='store_true', help='Use cached model')
              parser.add_argument('--num_iterations', type=int, help='Number of iterations')
              parser.add_argument('--model_url', help='Model URL')
              parser.add_argument('--dataset_name', help='Dataset name')
              parser.add_argument('--training_script', help='Path to training script')
              parser.add_argument('--hyperparameters', help='Hyperparameters for batch job (in JSON format)')
              parser.add_argument('--project_id', help='Project ID')

              args = parser.parse_args()
          "
        `)

        // The second response is ignored as there is no change in the diff
        const secondResponse = results[1] as SuggestedPredictionResult
        expect(secondResponse.type).toBe('ignored')
    })
})
