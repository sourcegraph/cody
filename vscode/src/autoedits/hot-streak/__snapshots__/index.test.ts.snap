// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processHotStreakResponses > does emit hot streaks when enabled 1`] = `
"
Code to Rewrite:

export function isEvenOrOdd(numberToChange: number): boolean {
    // Check if target is 0
    if (numberToChange === 0) {
        return true
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }

            
Code to Rewrite:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    // Check if target is 1
    if (numberToChange === 1) {
        return false
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }

            
Code to Rewrite:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    // Check if target is 2
    if (numberToChange === 2) {
        return true
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }

            
Code to Rewrite:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    // Check if target is 3
    if (numberToChange === 3) {
        return false
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }

            
Code to Rewrite:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }
    // Check if target is 4
    if (numberToChange === 4) {
        return true
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }
    if (target === 4) {
        return true
    }

            
Code to Rewrite:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }
    if (target === 4) {
        return true
    }
    // Check if target is 5
    if (numberToChange === 5) {
        return false
    }

Prediction:

export function isEvenOrOdd(target: number): boolean {
    if (target === 0) {
        return true
    }
    if (target === 1) {
        return false
    }
    if (target === 2) {
        return true
    }
    if (target === 3) {
        return false
    }
    if (target === 4) {
        return true
    }
    if (target === 5) {
        return false
    }

            "
`;
